### PR TITLE
meta, util: skip unmarshal unneeded JSON fields

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #55324

Problem Summary:

### What changed and how does it work?

manually process JSON tokens, to speed up the case where we only need a part of fields in JSON object

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

`fastjson` is the function that this PR uses. Also check the speed with `github.com/go-json-experiment/json v0.0.0-20240524174822-2d9f40f7385b`, which has the title `jsonv2`. Our tailored function still has better performance

```
goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/pkg/meta
cpu: Intel(R) Core(TM) i7-10710U CPU @ 1.10GHz
BenchmarkTableNameInfo
BenchmarkTableNameInfo/narrow-json
BenchmarkTableNameInfo/narrow-json-12         	   54498	     20050 ns/op
BenchmarkTableNameInfo/narrow-jsonv2
BenchmarkTableNameInfo/narrow-jsonv2-12       	   69752	     14474 ns/op
BenchmarkTableNameInfo/narrow-fastjson
BenchmarkTableNameInfo/narrow-fastjson-12     	  136118	     11306 ns/op
BenchmarkTableNameInfo/wide-json
BenchmarkTableNameInfo/wide-json-12           	   11055	     96475 ns/op
BenchmarkTableNameInfo/wide-jsonv2
BenchmarkTableNameInfo/wide-jsonv2-12         	   18374	     56001 ns/op
BenchmarkTableNameInfo/wide-fastjson
BenchmarkTableNameInfo/wide-fastjson-12       	  137173	     10351 ns/op
```

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by D3Hunter** [comment]
why https://github.com/pingcap/tidb/issues/49370 is refered, we don't check table existence from store now when create table.? see https://github.com/pingcap/tidb/pull/53882

what's the time takes after this pr for scenario in https://github.com/pingcap/tidb/issues/55324?

**Comment by lance6716** [comment]
> why #49370 is refered

I link that issue because this PR optimises more than https://github.com/pingcap/tidb/pull/49371, which is the fix PR for that issue. #49371 uses {"id": ..., "name": ...} rather than a full TableInfo to speed up. I will create a dedicated issue to link.

> what's the time takes after this pr for scenario in #55324?

#55324 requires more change, it will be implemented after this PR merged.

File: pkg/util/fastjson/extract.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+		return io.ErrUnexpectedEOF
+	}
+	return err
+}
+
+// next iterates and returns the top-level key and value as token. The odd
+// invocations return key tokens, the even invocations return value tokens. If
+// the value is a JSON array or object, all tokens belongs to the value will be
+// returned.
+func (i *topLevelJSONTokenIter) next(discard bool) ([]json.Token, error) {
```
Comment (D3Hunter) [inline pkg/util/fastjson/extract.go:48]: maybe split it into 2 methods to better fit what this struct named as
- nextKey
- readOrDiscardValue

Comment (lance6716) [reply pkg/util/fastjson/extract.go:73]: https://github.com/pingcap/tidb/pull/55304/commits/80ec0cb88843796d33da48cc20f3ddb98a383278

**Review by D3Hunter — APPROVED** [review]

**Review by tangenta — APPROVED** [review]

**Pull request merged**